frontend/src/modules/placement/classes.ts

```diff
@@ -0,0 +1,874 @@
+            }
+        }
+        return results
+    }
+
+
+    toString() {
+        return `RealContainerParentClass(name=${this.name})`
+    }
+
+    #resolveSamplePlacementIdentifier(placedSample: SamplePlacementIdentifier): SamplePlacement {
+        const cell = this.getCell(placedSample.cell)
+        const sample = this.placement.getSample(placedSample.sample)
+        return new SamplePlacement(this.context, { sample, cell })
+    }
+
+    // TODO: use dictionary parameter
+    getPlacements(onlySelected = false): SamplePlacement[] {
+        return Object.values(this.state.cells)
+            .flatMap((cell) =>
+                Object.keys(cell.samples).map((id) => new SamplePlacement(this.context, {
+                    sample: this.placement.getSample({ id: Number(id) }),
+                    cell: this.placement.getCell(cell)
+                }))
+            )
+            .filter((id) => onlySelected ? this.isPlacementSelected(id) : true)
+    }
+
+    // TODO: use dictionary parameter
+    getSortedPlacements(onlySelected = false): SamplePlacement[] {
+        return this.getPlacements(onlySelected).sort((a, b) => this.compareSamples(a, b))
+    }
+
+    sameContainerAs(containerID: ParentContainerIdentifier) {
+        return this.name === containerID.name
+    }
+
+    get name() {
+        return this.state.name
+    }
+    get spec() {
+        return this.state.spec
+    }
+
+    rawIdentifier(): RealParentContainerIdentifier {
+        return {
+            name: this.state.name
+        }
+    }
+}
+
+export class TubesWithoutParentClass extends PlacementObject {
+    readonly state: TubesWithoutParentContainerState
+
+    constructor(context: PlacementContext) {
+        super(context)
+        const state = context.placementState.tubesWithoutParentContainer
+        if (!state) {
+            throw new Error(`Tubes without parent container not found in state`)
+        }
+        this.state = state
+    }
+
+    isSampleSelected(sampleID: SampleIdentifier) {
+        return this.#getSampleEntry(sampleID).selected
+    }
+    selectSample(sampleID: SampleIdentifier) {
+        if (sampleID.id in this.state.samples) {
+            this.#getSampleEntry(sampleID).selected = true
+        }
+    }
+    unSelectSample(sampleID: SampleIdentifier) {
+        if (sampleID.id in this.state.samples) {
+            this.#getSampleEntry(sampleID).selected = false
+        }
+    }
+    toggleSelections(...sampleIDs: SampleIdentifier[]) {
+        const allSelected = sampleIDs.every(s => this.isSampleSelected(s))
+        if (allSelected) {
+            for (const sampleID of sampleIDs) {
+                this.unSelectSample(sampleID)
+            }
+        } else {
+            for (const sampleID of sampleIDs) {
+                this.selectSample(sampleID)
+            }
+        }
+    }
+
+    compareSamples(a: Pick<SamplePlacementIdentifier, 'sample'>, b: Pick<SamplePlacementIdentifier, 'sample'>) {
```

Comment:
I am not convinced you need more than one definition of this compareSample algorithm... There is non negligible risk here you will change one and forget the other ones ... the logic is the same as far as i can see... If you want to use the object inheritance you have here you could have your base class method do the basic sorting logic and have the derived classes prepare the data to call it properly using a super call.

Reply:
Ignore the object inheritance part here ... Would be too much of a rework.